stdlib/3.4/statistics.pyi

```diff
@@ -0,0 +1,44 @@
+# Stubs for statistics
+# See: http://docs.python.org/3/library/statistics.html
+
+import sys
+from typing import Iterable, Iterator, Sequence, TypeVar, Union, overload
+from decimal import Decimal
+
+# Note: according to the docs, statistics only explicitly supports
+# int, float, Decimal, and Fraction. Other types within the numeric
+# tower are not supported. It also states mixing together different
+# types results in undefined behavior, so the type signatures below
+# deliberately enforce that numeric types may not be mixed.
+#
+# TODO: Once either https://github.com/python/typeshed/pull/545 or
+# https://github.com/python/typeshed/pull/94 is accepted, add support
+# for fractions.
+_TNum = TypeVar('_TNum', int, float, Decimal)
+_T = TypeVar('_T')
+
+def mean(data: Union[Iterator[_TNum], Sequence[_TNum]]) -> _TNum: ...
```

Comment:
I think the worry is mostly about float vs. Decimal, where the user has to
decide. int can be mixed with each of the others. But probably we can start
by making the type variable allow float and Decimal only.

On Wed, Sep 14, 2016 at 8:36 AM, Michael Lee <notifications@github.com>
wrote:

> In stdlib/3.4/statistics.pyi
> <https://github.com/python/typeshed/pull/546#discussion_r78774231>:
>
> > +from typing import Iterable, Iterator, Sequence, TypeVar, Union, overload
> > +from decimal import Decimal
> > +
> > +# Note: according to the docs, statistics only explicitly supports
> > +# int, float, Decimal, and Fraction. Other types within the numeric
> > +# tower are not supported. It also states mixing together different
> > +# types results in undefined behavior, so the type signatures below
> > +# deliberately enforce that numeric types may not be mixed.
> > +#
> > +# TODO: Once either https://github.com/python/typeshed/pull/545 or
> > +# https://github.com/python/typeshed/pull/94 is accepted, add support
> > +# for fractions.
> > +_TNum = TypeVar('_TNum', int, float, Decimal)
> > +_T = TypeVar('_T')
> > +
> > +def mean(data: Union[Iterator[_TNum], Sequence[_TNum]]) -> _TNum: ...
>
> Oh, hmm -- when I did mean([1, 2, 3]), I got back an int, but after some
> poking, that turned out to be only because 1 + 2 + 3 is actually evenly
> divisible by 3.
>
> Perhaps I could get rid of int within the TypeVar and rely on mypy to
> understand that ints should be auto-promoted to float?
>
> Regarding mixing ints with Decimals/mixing floats with Decimals: the documentation
> for statistics <https://docs.python.org/3/library/statistics.html> stated
> that mixing together different types is "undefined and
> implementation-dependent", and even recommends running map(float,
> input_data) first if you have mixed numeric types.
>
> So, if mixing together things like Decimal and int is undefined, then I
> think it's ok and actually for the best if typeshed forbids that altogether?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/python/typeshed/pull/546/files/98161406c69d8f31f4b848dffb889c04a5e49098#r78774231>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACwrMiPTW4ogrthvoNIOaKJ3XE9k9XwIks5qqBSTgaJpZM4J8bk1>
> .
>



-- 
--Guido van Rossum (python.org/~guido)